cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -0,0 +1,890 @@
+
+  Status DoubleSize() { return Base::DoubleTableSize(); }
+
+  Status Append(const ArrayData& input) override { return Base::Append(input); }
+
+  Status Flush(std::vector<Datum>* out) override {
+    // No-op
+    return Status::OK();
+  }
+};
+
+// ----------------------------------------------------------------------
+// Dictionary encode implementation
+
+template <typename Type>
+class DictEncodeImpl : public HashTableKernel<Type, DictEncodeImpl<Type>> {
+ public:
+  static constexpr bool allow_expand = true;
+  using Base = HashTableKernel<Type, DictEncodeImpl>;
+
+  DictEncodeImpl(const std::shared_ptr<DataType>& type, MemoryPool* pool)
+      : Base(type, pool), indices_builder_(pool) {}
+
+  Status Reserve(const int64_t length) { return indices_builder_.Reserve(length); }
+
+  void ObserveNull() { indices_builder_.UnsafeAppendToBitmap(false); }
+
+  void ObserveFound(const hash_slot_t slot) { indices_builder_.UnsafeAppend(slot); }
+
+  void ObserveNotFound(const hash_slot_t slot) { return ObserveFound(slot); }
+
+  Status DoubleSize() { return Base::DoubleTableSize(); }
+
+  Status Flush(std::vector<Datum>* out) override {
+    std::shared_ptr<ArrayData> result;
+    RETURN_NOT_OK(indices_builder_.FinishInternal(&result));
+    out->push_back(Datum(result));
+    return Status::OK();
+  }
+
+  using Base::Append;
+
+ private:
+  Int32Builder indices_builder_;
+};
+
+// ----------------------------------------------------------------------
+// Kernel wrapper for generic hash table kernels
+
+class HashKernelImpl : public HashKernel {
+ public:
+  explicit HashKernelImpl(std::unique_ptr<HashTable> hasher)
+      : hasher_(std::move(hasher)) {}
+
+  Status Call(FunctionContext* ctx, const ArrayData& input,
+              std::vector<Datum>* out) override {
+    RETURN_NOT_OK(Append(ctx, input));
+    return Flush(out);
+  }
+
+  Status Append(FunctionContext* ctx, const ArrayData& input) override {
+    std::lock_guard<std::mutex> guard(lock_);
+    try {
+      RETURN_NOT_OK(hasher_->Append(input));
+    } catch (const HashException& e) {
+      return Status(e.code(), e.what());
+    }
+    return Status::OK();
+  }
+
+  Status Flush(std::vector<Datum>* out) override { return hasher_->Flush(out); }
+
+  Status GetDictionary(std::shared_ptr<ArrayData>* out) override {
+    return hasher_->GetDictionary(out);
+  }
+
+ private:
+  std::mutex lock_;
+  std::unique_ptr<HashTable> hasher_;
+};
+
+}  // namespace
+
+Status GetUniqueKernel(FunctionContext* ctx, const std::shared_ptr<DataType>& type,
+                       std::unique_ptr<HashKernel>* out) {
+  std::unique_ptr<HashTable> hasher;
+
+#define UNIQUE_CASE(InType)                                         \
+  case InType::type_id:                                             \
+    hasher.reset(new UniqueImpl<InType>(type, ctx->memory_pool())); \
+    break
+
+  switch (type->id()) {
+    UNIQUE_CASE(NullType);
+    // UNIQUE_CASE(BooleanType);
```

Comment:
Still need to implement for `BooleanType`. This can be simplified since the hash table can have at most 2 (or 3, if you count nulls) values